### PR TITLE
[Hub Generated] Review request for Microsoft.DBforPostgreSQL to add version preview/2020-11-05-preview

### PR DESCRIPTION
**Pull request by tezhangmsvan**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/tezhangmsvan/azure-rest-api-specs/dev-postgresql-Microsoft.DBforPostgreSQL-2020-11-05-preview-tezhang...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [x] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [x] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [x] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [x] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [x] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by chiragg4u** [comment]
Please go through the contribution checklist for ARM API Review added as part of the review and mark them completed. It will help you fix most of the issues and improve the quality of swagger and help close the PR faster. This is a requirement before review can be continued.

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+        "targetDBServerMetadata": {
+          "$ref": "#/definitions/DBServerMetadata"
+        },
+        "sourceDBServerResourceId": {
+          "type": "string"
+        },
+        "secretParameters": {
+          "$ref": "#/definitions/MigrationSecretParameters"
+        },
+        "targetDBServerResourceId": {
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:314]: why are source and target necessary?  Is one not implied by the parent resource this migration is created under?

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+        {
+          "$ref": "../../../../../common-types/resource-management/v1/types.json#/definitions/TrackedResource"
+        }
+      ],
+      "description": "Represents a migration resource."
+    },
+    "MigrationResourceProperties": {
+      "type": "object",
+      "properties": {
+        "migrationId": {
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:292]: please indicate in the schema what fields are readonly and what is required as input

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
+      "flow": "implicit",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{targetDBServerSubscriptionId}/resourceGroups/{targetDBServerResourceGroupName}/providers/Microsoft.DBforPostgreSQL/flexibleServers/{targetDBServerName}/migrations/{migrationId}": {
+      "put": {
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:38]: how long does the migration take?  Should this not be an async operation?

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+          "type": "string"
+        },
+        "location": {
+          "maxLength": 128,
+          "type": "string"
+        }
+      },
+      "description": "Migration resource group."
+    },
+    "MigrationSecretParameters": {
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:417]: what are these used for and why can't Managed Resource Identity be used instead?

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 5]
```diff
@@ -0,0 +1,655 @@
...
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{targetDBServerSubscriptionId}/resourceGroups/{targetDBServerResourceGroupName}/providers/Microsoft.DBforPostgreSQL/flexibleServers/{targetDBServerName}/migrations": {
+      "get": {
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:227]: this should be marked x-ms-pageable

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+            "$ref": "#/parameters/TargetDBServerSubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/TargetDBServerResourceGroupNameParameter"
+          },
+          {
+            "$ref": "#/parameters/TargetDBServerNameParameter"
+          },
+          {
+            "name": "migrationListFilter",
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:249]: any sort of query filtering should supply a rich description detailing what is supported

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+          "type": "string"
+        },
+        "dBsToMigrate": {
+          "maxItems": 8,
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "migrationResourceGroup": {
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:327]: what is this resource group used for?

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+          "format": "uuid",
+          "type": "string"
+        },
+        "migrationDetailsLevel": {
+          "$ref": "#/definitions/MigrationDetailsLevel"
+        },
+        "currentStatus": {
+          "$ref": "#/definitions/MigrationStatus"
+        },
+        "sourceDBServerMetadata": {
```
[the remark this answers is not in the document]
Comment (tezhangmsvan) [reply specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:302]: This is generated on the  server side. Will mark as readonly

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+          "type": "boolean"
+        },
+        "overwriteDBsInTarget": {
+          "type": "boolean"
+        },
+        "migrationWindowStartTimeInUtc": {
+          "format": "date-time",
+          "type": "string"
+        },
+        "startDataMigration": {
```
[the remark this answers is not in the document]
Comment (tezhangmsvan) [reply specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:340]: Good point. Only startDataMigration and Cancel can interfere here, removing cancel. Cancel will be handled in delete API. Other flags can be set independently.

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json (hunk 4 above)
Comment (tezhangmsvan) [reply specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:417]: This is a private preview version. Will move to Managed identity in public preview.

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json (hunk 7 above)
Comment (tezhangmsvan) [reply specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json:327]: When the users sets startDataMigration flag to true, we will provision a Database Migration Service(DMS) under the hood an d DMS will do the heavy lifting of the migration. This resource group is used to for provisioning DMS.

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json (hunk 6 above)
Comment (tezhangmsvan) [reply specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2020-11-05-preview/Migrations.json]: Changed to reference a parameter

**Comment by tezhangmsvan** [comment]
@pilor resolved all comments and ready for another iteration of review. Thanks!

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json [hunk 10]
```diff
@@ -0,0 +1,707 @@
...
+          },
+          {
+            "$ref": "#/parameters/migrationNameParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/MigrationResource"
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json:223]: missed this on the first iteration.  Deletes should not return a body.

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json [hunk 11]
```diff
@@ -0,0 +1,707 @@
...
+          },
+          {
+            "$ref": "#/parameters/TargetDBServerResourceGroupNameParameter"
+          },
+          {
+            "$ref": "#/parameters/TargetDBServerNameParameter"
+          },
+          {
+            "name": "parameters",
+            "in": "body",
```
Comment (pilor) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json:265]: GETs cannot take a body.  Any filtering should be done in the URI.  The initial iteration had it that way right?  All that was needed was a better description of the filtering that is supported

**Comment by tezhangmsvan** [comment]
@pilor resolved comments. Thanks

**Review by Rajesh1Ganesan — APPROVED** [review]

File: specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json [hunk 12]
```diff
@@ -0,0 +1,818 @@
...
+        },
+        "message": {
+          "type": "string",
+          "description": "Migration name availability message.",
+          "readOnly": true
+        }
+      },
+      "description": "Represents a migration name's availability."
+    },
+    "CloudError": {
```
Comment (jianyexi) [inline specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json:758]: Could you use  the ErrorResponse in v2/types.json instead ?

Comment (tezhangmsvan) [reply specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2021-05-31-privatepreview/Migrations.json:762]: We use this object across our api's see Microsoft.DBforPostgreSQL/stable/2021-06-01/postgresql.json

**Comment by tezhangmsvan** [comment]
@jianyexi the check https://github.com/Azure/azure-rest-api-specs/pull/13555/checks?check_run_id=2835376627 
 is giving me:

The schema 'StorageEditionCapability' is considered an object but without a 'type:object', please add the missing 'type:object'.Location:&nbsp;Microsoft.DBforPostgreSQL/preview/2021-06-01-preview/postgresql.json#L1161
--


But if I add the type:object, it will complain that is a breaking change because the type changed. How do I resolve this?

**Comment by jianyexi** [comment]
@tezhangmsvan  
There is only one breaking change you needs to get approval:
The new version is missing a path that was found in the old version. Was path '/providers/Microsoft.DBForPostgreSql/getPrivateDnsZoneSuffix' removed or restructured?
Old: Microsoft.DBforPostgreSQL/stable/2021-06-01/PrivateDnsZone.json#L37:5

**Comment by jianyexi** [comment]
@tezhangmsvan  why you add a private preview version into the public spec repo ?

**Comment by tezhangmsvan** [comment]
> @tezhangmsvan why you add a private preview version into the public spec repo ?

image

**Comment by tezhangmsvan** [comment]
> > @tezhangmsvan why you add a private preview version into the public spec repo ?
> 
> image

I think private means different things in the private repo and a private preview version. A private repo means internal to us and private preview means private preview for a customer if I understand correctly,

**Comment by tezhangmsvan** [comment]
@jianyexi Validations passed. Can you approve? Thanks.

**Comment by jianyexi** [comment]
> @jianyexi Jianye Xi FTE Validations passed. Can you approve? Thanks.
Since there are cross version breaking changes, could you follow the breaking change review process to get approval
https://github.com/Azure/azure-rest-api-specs/pull/13555#issuecomment-861869913

**Comment by tezhangmsvan** [comment]
> > @jianyexi Jianye Xi FTE Validations passed. Can you approve? Thanks.
> > Since there are cross version breaking changes, could you follow the breaking change review process to get approval
> > [#13555 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/13555#issuecomment-861869913)

So when I reference old API files, it tells that "XX  is considered an object but without a 'type:object', please add the missing 'type:object'". Then if I add them, it's considered a type chang and thus a breaking change. How do I resolve this?

**Comment by jianyexi** [comment]
> > > @jianyexi
> > >             
> > >             Jianye Xi
> > >             FTE Jianye Xi FTE Validations passed. Can you approve? Thanks.
> > > Since there are cross version breaking changes, could you follow the breaking change review process to get approval
> > > [#13555 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/13555#issuecomment-861869913)
> 
> So when I reference old API files, it tells that "XX is considered an object but without a 'type:object', please add the missing 'type:object'". Then if I add them, it's considered a type chang and thus a breaking change. How do I resolve this?
Just add the missing 'type:object' , it's not breaking change now

**Comment by jianyexi** [comment]
I mean the below breaking change need get approval, not the breaking change caused by adding missing type object:
image

**Comment by tezhangmsvan** [comment]
> I mean the below breaking change need get approval, not the breaking change caused by adding missing type object:
> image

That was outdated with my latest changes. Please see my latest changes and checks.

**Comment by tezhangmsvan** [comment]
@jianyexi Swagger LintDiff  complains about errors in our exsiting API. How do I resolve this? If I edit the exist APIs by adding the object type, it'll tell me it's a breaking change.

**Comment by tezhangmsvan** [comment]
@jianye operations get was clearly implemented, why is it complaining?
Operations API must be implemented for '/providers/Microsoft.DBForPostgreSQL/operations'.
Location: Microsoft.DBforPostgreSQL/preview/2021-06-15-privatepreview/postgresql.json#L36

**Comment by tezhangmsvan** [comment]
@jianyexi looks all are passing. Can you approve?

**Comment by jianyexi** [comment]
> @jianyexi Jianye Xi FTE looks all are passing. Can you approve?
> 
> So I added the "type : obejct" to conform, but then the cross version breaking changes consider this a type change and breaking? How do I resolve this?

It's not breaking change now, are you ready for merging ?

**Comment by tezhangmsvan** [comment]
> > @jianyexi Jianye Xi FTE looks all are passing. Can you approve?
> > So I added the "type : obejct" to conform, but then the cross version breaking changes consider this a type change and breaking? How do I resolve this?
> 
> It's not breaking change now, are you ready for merging ?

Yes, we are ready to merge. Thanks.

**Comment by tezhangmsvan** [comment]
@jianyexi ready to merge. Thanks

**Pull request merged**